src/goto-instrument/reachability_slicer.cpp

```diff
@@ -54,11 +54,49 @@ void reachability_slicert::fixedpoint_to_assertions(
   const is_threadedt &is_threaded,
   slicing_criteriont &criterion)
 {
-  std::vector<cfgt::node_indext> src = get_sources(is_threaded, criterion);
+  // Stack entries are pairs of node indices and a boolean flag that indicates
+  // whether the function's callsite is known, in which case it has already
+  // been placed on the stack and function heads are a no-op.
 
-  std::vector<cfgt::node_indext> reachable = cfg.get_reachable(src, false);
-  for(const auto index : reachable)
-    cfg[index].reaches_assertion = true;
+  std::vector<std::pair<cfgt::node_indext, bool>> stack;
+  std::vector<cfgt::node_indext> sources = get_sources(is_threaded, criterion);
+  for(const auto source : sources)
+    stack.emplace_back(source, false);
+
+  while(!stack.empty())
+  {
+    auto index = stack.back().first;
+    auto callsite_is_known = stack.back().second;
+    stack.pop_back();
+
+    auto &node = cfg[index];
+    if(node.reaches_assertion)
+      continue;
+    node.reaches_assertion = true;
+
+    for(const auto &edge : node.in)
+    {
+      const auto &pred_node = cfg[edge.first];
+
+      if(pred_node.PC->is_end_function())
+      {
+        stack.emplace_back(edge.first, true);
+        stack.emplace_back(
+          cfg.entry_map[std::prev(node.PC)], callsite_is_known);
```

Comment:
The use of `std::prev` is neither safe (as the Travis failure shows) nor sound: an instruction may have multiple incoming edges on the CFG. I believe you will need to go via the incoming edges on the CFG rather than just taking the predecessor in the list of goto program instructions.

Reply:
It may, but that imprecision is exactly what I want to eliminate. If we entered this function by way of a particular callsite (indicated by `callsite_is_known`) then we should not further explore all possible callers, only the one that's actually reachable. 

The problem was simply taking `std::prev` of the head of a function -- I've fixed this using `std::next(pred_it)` instead.